src/Storages/Streaming/StreamingStoreSourceBase.cpp

```diff
@@ -138,8 +138,18 @@ Chunk StreamingStoreSourceBase::generate()
     if (isCancelled())
         return {};
 
-    if (auto * current = ckpt_ctx.exchange(nullptr, std::memory_order_relaxed); current)
-        return doCheckpoint(CheckpointContextPtr{current});
+    CheckpointContextPtr current;
+    if (has_ckpt_request)
+    {
+        {
+            std::scoped_lock lock(ckpt_mutex);
+            current.swap(ckpt_ctx);
+        }
+        has_ckpt_request = false;
```

Comment:
We will need set it under ckpt_mutex held.

Reply:
why? Does Atomic Value also need a mutex guard?